### ISSUE TITLE
gearmand hanging and non responsive after OMD 3.30 upgrade

### ISSUE THREAD

**Issue by infraweavers** [ConSol/omd#107]
After upgrading to OMD 3.30 from 3.10 we find that gearmand hangs and is no longer able to be connected to, meaning all the checks timeout. We use mod_gearman to provide `dupserver` which we then submit the results to our secondary OMD box.

omd restart gearmand fails to kill the process and the only solution is to kill -9 gearmand. 

![cantstop](https://user-images.githubusercontent.com/36475663/89882935-6cded200-dbbf-11ea-9578-b43ca5cf5f09.png)

We have core files of gearmand when it's in this state and ran strace against it, which showed no activity occurring in the process.

![strace_gdbattached](https://user-images.githubusercontent.com/36475663/89882917-681a1e00-dbbf-11ea-838e-c55ae7d404b4.png)

The error from running gearman_top when its hung is:
```
failed to connect to address localhost and port 4730: Interrupted system call.
```
![Gearman_top](https://user-images.githubusercontent.com/36475663/89882957-749e7680-dbbf-11ea-8f80-f4d52be6cfd7.png)

which is very unusual; if gearmand is stopped or unavailable, connection refused is the error.

We think this probably related to the upgrade from gearmand 0.33 that we think we occurred in OMD 3.20? Where is the best place to discuss the issue, is here fine?

**Comment by sni** [comment]
the place to discuss this is fine. I see this on one box occasionally as well but haven't found time to dig deeper into it.

**Comment by infraweavers** [comment]
We've tried to reproduce this in a lab by copying the naemon check configuration and all the plugins onto a test OMD 3.30 install, the problem has not reocurred in the lab; so we're now wondering if it's somethign specific with the output of one of our checks or similiar. So we've built another OMD 3.30, this time in the same environment as the ones we upgraded previous and it's running all the same checks in parallel to OMD3.10; the only difference in configuration between the 2 is Keepalived and Postfix are both disabled on the OMD3.30; and the `dupserver` is commented out on OMD3.30 at the moment. If we can't reproduce the problem in a few hours, we'll create another OMD3.30 and configure the `dupserver` so it is identical to the upgraded pair we had the issue on above.

**Comment by infraweavers** [comment]
This has just been reproduced now, so presumably the specific checks we're running are important.

**Comment by sni** [comment]
thats strange, because mod-gearman encrypts and converts the crypted payload with base64 to avoid such problems. Except for the `uniq` identifier. So it might have to do something with a specific host or servicename.

**Comment by infraweavers** [comment]
We've also just had this scenario (`-1 jobs`) appear whilst trying to narrow it down:

![image](https://user-images.githubusercontent.com/36475663/90020868-595b6600-dca8-11ea-80e0-8fa82013e34a.png)


Looks like `gearadmin --status` wraps the integer:
![image](https://user-images.githubusercontent.com/36475663/90021226-ddade900-dca8-11ea-8304-dca9a689168c.png)

**Comment by infraweavers** [comment]
@sni we use gcore to get the coredumps, we think we need a debug build of gearmand with symbols to feed into `gdb` do you know how we can build one or get the required files?

**Comment by infraweavers** [comment]
We think we got this working, looks like just building gearmand from OMD 3.30's release commit was enough:

![image](https://user-images.githubusercontent.com/36475663/90114675-f5d64480-dd4a-11ea-8b06-b14538e870cc.png)

**Comment by sni** [comment]
yeah, you can simply run `make` from the `packages/gearmand` folder. Eventually you need to run `./configure` from the base folder to install all build dependencies. And you might want to adjust some cflags to increase debug symbols if necessary.

Seems that gearmands internal state gets confused somehow.

**Comment by infraweavers** [comment]
We just got it reproduced again, however this time we were tailing the gearmand log, 
![image](https://user-images.githubusercontent.com/36475663/90123870-66835e00-dd57-11ea-8b31-0230108f2307.png)

We saw the gearman_top have -1 in the check_results, that correllated with the log appearing; shortly after the gearmand stopped responding completely.

**Comment by infraweavers** [comment]
We managed to catch and attach to this happening, and it's basically in an infinite loop between:
https://github.com/gearman/gearmand/blob/master/libgearman-server/worker.cc#L124-L131

**Comment by sni** [comment]
great debugging, i guess its time to get the gearmand people involved.

**Comment by infraweavers** [comment]
We've opened an issue here: https://github.com/gearman/gearmand/issues/301 we think that's the right place for it

**Comment by infraweavers** [comment]
@sni We do use pnp_gearman_worker for our graphing on the affected environment; we'll make this change (https://github.com/ConSol/omd/commit/dd8b990c6b033a74e6c86a69b25dcfbd21e6b7a5) on our test OMD 3.30 box and see if it makes any difference to it.

**Comment by sni** [comment]
are you using pnp in gearman mode?

**Comment by infraweavers** [comment]
Yes

**Comment by sni** [comment]
ok, then it might be the same issue.

**Comment by infraweavers** [comment]
@sni We patched our test OMD 3.30 box on the 28th and have it running since. It doesn't seem to have locked up with `interrupted system call` in the few days we've been running it, whereas before it would almost certainly have done it once by now. So it seems like https://github.com/ConSol/omd/commit/dd8b990c6b033a74e6c86a69b25dcfbd21e6b7a5 may have worked around the gearmand bug in our environment. We'll keep testing it for a bit and see if anything changes.

**Comment by sni** [comment]
same here, no outage since then. Looks good.

**Comment by infraweavers** [comment]
@sni just tried promoting one of our actual production boxes up to 3.30 and applied https://github.com/ConSol/omd/commit/dd8b990c6b033a74e6c86a69b25dcfbd21e6b7a5, unfortunately it just broke again with the exact same behaviour as before, so it doesn't look like it's fixed it for us :(

**Comment by sni** [comment]
still no issues here since with the latest nightly

**Comment by infraweavers** [comment]
Interesting, we're testing with the nightly rather than manually applying https://github.com/ConSol/omd/commit/dd8b990c6b033a74e6c86a69b25dcfbd21e6b7a5 to see if we get it again

**Comment by infraweavers** [comment]
@sni Just checking through the mod_gearman code for anything looking "weird"; noticed this: https://github.com/sni/mod_gearman/blob/master/common/gearman_utils.c#L97. Is that *definitely* the connection to gearmand network timeout and not a job timeout (i.e. `CAN_DO_TIMEOUT` in gearman parlance) ?

**Comment by infraweavers** [comment]
Nightly has been running for a few hours now and we've not seen the problem yet

**Comment by infraweavers** [comment]
We're going to upgrade our secondary at our first site to the nightly and see if the stability remains. It's looking pretty good so far, however!

**Comment by infraweavers** [comment]
Looking stable so far, there does seem to be a pause in the behaviour of checks when the `dupserver` is unavailable; however that is a totally separate issue from this. The good news is, it does seem to be fixed so far...

**Comment by infraweavers** [comment]
Just reporting in, this has been stable for another week; I'm pretty convinced this has worked around the gearmand problem now :)

**Comment by sni** [comment]
Thats great to hear. Will close this then
